internal/cache/file/cache_handler_test.go

```diff
@@ -523,192 +643,330 @@ func (chrT *cacheHandlerTest) Test_GetCacheHandle_ConcurrentDifferentFiles() {
+			file, err := os.OpenFile(downloadPath, os.O_RDONLY, 0600)
+			require.NoError(t, err)
+
+			err = chTestArgs.cacheHandler.InvalidateCache(minObject.Name, chTestArgs.bucket.Name())
+
+			assert.ErrorIs(t, err, tc.expectedCInvalidateCacheErr)
+			// Reading from the open file handle should fail as the file is truncated.
+			_, err = file.Read(buf)
+			assert.ErrorIs(t, err, tc.expectedCacheFileReadErr)
+		})
 	}
-	wg.Wait()
 }
 
-func (chrT *cacheHandlerTest) Test_InvalidateCache_ConcurrentDifferentFiles() {
-	wg := sync.WaitGroup{}
-
-	InvalidateCacheTestFun := func(index int) {
-		defer wg.Done()
-		objName := "object" + strconv.Itoa(index)
-		objContent := "object content: content#" + strconv.Itoa(index)
-		minObj := chrT.getMinObject(objName, []byte(objContent))
-
-		err := chrT.cacheHandler.InvalidateCache(minObj.Name, chrT.bucket.Name())
-
-		assert.NoError(chrT.T(), err)
-		assert.Nil(chrT.T(), chrT.jobManager.GetJob(objName, chrT.bucket.Name()))
-		assert.False(chrT.T(), chrT.isEntryInFileInfoCache(objName, chrT.bucket.Name()))
+func Test_InvalidateCache_ConcurrentSameFile(t *testing.T) {
+	tbl := []struct {
+		name            string
+		fileCacheConfig config.FileCacheConfig
+		cacheDir        string
+	}{
+		{
+			name:            "Non parallel downloads",
+			fileCacheConfig: config.FileCacheConfig{EnableCRC: true},
+			cacheDir:        path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+		},
+		{
+			name: "Parallel downloads",
+			fileCacheConfig: config.FileCacheConfig{EnableCRC: true, EnableParallelDownloads: true,
+				ParallelDownloadsPerFile: 1, MaxParallelDownloads: 20, DownloadChunkSizeMB: 3},
+			cacheDir: path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+		},
 	}
-
-	// Start concurrent GetCacheHandle()
-	for i := 0; i < 5; i++ {
-		wg.Add(1)
-		go InvalidateCacheTestFun(i)
+	for _, tc := range tbl {
+		t.Run(tc.name, func(t *testing.T) {
+			chTestArgs := initializeCacheHandlerTestArgs(t, &tc.fileCacheConfig, tc.cacheDir)
+			existingJob := getDownloadJobForTestObject(t, chTestArgs)
+			require.Equal(t, downloader.NotStarted, existingJob.GetStatus().Name)
+			require.True(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
+			wg := sync.WaitGroup{}
+			InvalidateCacheTestFun := func(t *testing.T) {
+				defer wg.Done()
+
+				err := chTestArgs.cacheHandler.InvalidateCache(chTestArgs.object.Name, chTestArgs.bucket.Name())
+
+				assert.NoError(t, err)
+				assert.NotNil(t, existingJob)
+				assert.Equal(t, downloader.Invalid, existingJob.GetStatus().Name)
+				assert.Equal(t, false, doesFileExist(t, chTestArgs.downloadPath))
+				// File info should also be removed.
+				assert.False(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
+			}
+
+			// Start concurrent GetCacheHandle()
+			for i := 0; i < 5; i++ {
+				wg.Add(1)
+				go InvalidateCacheTestFun(t)
+			}
```

Comment:
The function can access "t" from outer scope.

```suggestion
			InvalidateCacheTestFun := func() {
				defer wg.Done()

				err := chTestArgs.cacheHandler.InvalidateCache(chTestArgs.object.Name, chTestArgs.bucket.Name())

				assert.NoError(t, err)
				assert.NotNil(t, existingJob)
				assert.Equal(t, downloader.Invalid, existingJob.GetStatus().Name)
				assert.Equal(t, false, doesFileExist(t, chTestArgs.downloadPath))
				// File info should also be removed.
				assert.False(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
			}

			// Start concurrent GetCacheHandle()
			for i := 0; i < 5; i++ {
				wg.Add(1)
				go InvalidateCacheTestFun()
			}
```